pallets/wnodl/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+        WrappingInitiated(T::AccountId, BalanceOf<T>, EthAddress),
+
+        /// Wrapping Nodl is settles
+        /// parameters. [account's address on Nodle chain, amount of Nodl fund settled, Transaction hash on Ethereum main-net, reporting oracle's address]
+        WrappingSettled(T::AccountId, BalanceOf<T>, EthTxHash),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The balance of the account is not sufficient for the requested transaction
+        BalanceNotEnough,
+        /// The amount of fund to wrap should be between the pre-specified limits for the pallet.
+        FundNotWithinLimits,
+        /// We do not expect this error ever happen. But if it happened we would not allow it to mess with the correctness of the storage.
+        BalanceOverflow,
+        /// The customer is not known, whitelisted for this operation.
+        NotEligible,
+        /// Settlment is only possible for an amount equal or smaller than an initiated wrapping for a known customer
+        InvalidSettle,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Initiate wrapping an amount of Nodl into wnodl on Ethereum
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        pub fn initiate_wrapping(
+            origin: OriginFor<T>,
+            amount: BalanceOf<T>,
+            eth_dest: EthAddress,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            #[cfg(feature = "runtime-benchmarks")]
+            if let Some(whitelisted_callers) = WhitelistedCallers::<T>::get() {
+                ensure!(whitelisted_callers.contains(&who), Error::<T>::NotEligible);
+            } else {
+                ensure!(T::KnownCustomers::contains(&who), Error::<T>::NotEligible);
+            }
+            #[cfg(not(feature = "runtime-benchmarks"))]
+            ensure!(T::KnownCustomers::contains(&who), Error::<T>::NotEligible);
+
+            let free_balance = T::Currency::free_balance(&who);
+            let minimum_balance = T::Currency::minimum_balance();
+            ensure!(
+                free_balance >= amount.saturating_add(minimum_balance),
+                Error::<T>::BalanceNotEnough
+            );
+
+            let current_sum = TotalInitiated::<T>::get().unwrap_or_else(Zero::zero);
+            let total = current_sum
+                .checked_add(&amount)
+                .ok_or::<Error<T>>(Error::BalanceOverflow)?;
+
+            let balances = Balances::<T>::get(who.clone()).unwrap_or((Zero::zero(), Zero::zero()));
+            let total_for_origin = balances
+                .0
+                .checked_add(&amount)
+                .ok_or::<Error<T>>(Error::BalanceOverflow)?;
+
+            TotalInitiated::<T>::put(total);
+            Balances::<T>::mutate(who.clone(), |x| {
+                *x = Some((total_for_origin, balances.1));
+            });
```

Comment:
We probably want to `put` here to save one read since the value was read before

Reply:
Done